app/gui2/src/components/GraphEditor.vue

```diff
@@ -21,12 +23,27 @@ const viewportNode = ref<HTMLElement>()
 const navigator = useNavigator(viewportNode)
 const graphStore = useGraphStore()
 const projectStore = useProjectStore()
+const executionCtx = ref<Promise<Opt<ContextId>>>()
```

Comment:
Promises in `ref`s is usually not what you want, as it's a pain to work with. You usually want one of two other options:
- create a class that encapsulates the asynchronous nature of something,
- or assign `ref` value once async call is done (usually can be done using `computedAsync` function)

In this case, I think we could actually benefit from the first approach. An `ExecutionContext` class that knows its ID, and it implements its own async queue of pending operations. That way you can create and assign it synchronously. You can also make execution stack management to appear synchronous from the outside. The actual data responses can send appropriate events through `ObservableV2` interface (e.g. execution status, expression updates, visualization execution).